docs/source/design/user_messaging.rst

```diff
@@ -0,0 +1,128 @@
+records from a sub-process, these should be inspected individually in order to identify result
+values that could require re-annotation or status re-classification.
+
+For user messaging purposes, result records can also be enriched with additional human-readable
+information on the nature of the result, via the ``message`` key, and human-readable hints to
+the user, via the ``hints`` key. Both of these are rendered via the `UI Module`_.
+
+
+Exception handling
+------------------
+
+In general, **exceptions should be raised when there is no way to ignore or recover from
+the offending action**.
+
+More specifically, raise an exception when:
+
+1. A DataLad interface's parameter specifications are violated
+2. An additional requirement (beyond parameters) for the meaningful continuation of a
+   command, function, or process is not met
+
+It must be made clear to the user/caller what the exact cause of the exception
+is, given the context within which the user/caller triggered the action.
+This is achieved directly via a (re)raised exception, as opposed to logging messages or
+results records which could be ignored or unseen by the user.
+
+.. note::
+   In the case of a complex set of dependent actions it could be expensive to
+   confirm parameter violations. In such cases, initial sub-routines might already generate
+   result records that have to be inspected by the caller, and it could be practically better
+   to yield a result record (with ``status=[error|impossible]``) to communicate the failure.
+   It would then be up to the upstream caller to decide whether to specify
+   ``on_failure='ignore'`` or whether to inspect individual result records and turn them
+   into exceptions or not.
+
+
+Logging
+-------
+
+Logging provides developers with additional means to describe steps in a process,
+so as to **allow insight into the program flow during debugging** or analysis of e.g.
+usage patterns. Logging can be turned off externally, filtered, and redirected. Apart from
+the :ref:`log-level <chap_design_log_levels>` and message, it is not inspectable and
+cannot be used to control the logic or flow of a program.
+
+Importantly, logging should not be the primary user messaging method for command outcomes,
+Therefore:
+
+1. No interface should rely solely on logging for user communication
+2. Use logging for in-progress user communication via the mechanism for :ref:`progress reporting <chap_design_progress_reporting>`
+3. Use logging to inform debugging processes
+
+
+UI Module
+---------
+
+The :mod:`~datalad.ui` module provides the means to communicate information
+to the user in a user-interface-specific manner, e.g. via a console, dialog, or an iPython interface.
+Internally, all DataLad results processed by the result renderer are passed through the UI module.
+
+nd necessary to deliver message to the user 
+	is not "message" of the result record,  developers should let explicit user communication
```

Comment:
I feel like this is a copy-paste victim?

Reply:
LOL thanks for catching. indeed.

Reply:
fixed in fc28212154d8f595de6d81d1fdcbe22466a81be0